### PR TITLE
Repository Reorganization

### LINKED ISSUE #15: Update SpecTf architecture with latest version from GHG

### LINKED ISSUE #16: Formalize and integrate function/class imports

### PR DESCRIPTION
**Pull request by jakehlee**
Closes #16 
Closes #15 

## Changelog

1. ⚠️ The `cloud` dir is split into `spectf` and `spectf_cloud` e5e68d84a9a2613f5835e877c829ba969983833a
2. ⚠️ The `cloud` CLI entry point is renamed to `spectf_cloud` 
3. 🛠️ Tests fixed so that dummy labels are integers/longs for cross-entropy c44be85c66410619bec2c5a34ebceb80da43fbbc
4. 🎉 Updated model architecture 4245da9e7f18fc506963c28a82448f2cf08299ad
5. 🧹 Pylinted and added docstrings for `spectf`

### `cloud` bifrucation and `spectf_cloud` rename

This bifrucation reorients the repository into its two roles:
1. As a companion code repository for publications
2. As a importable package for future work

In anticipation of other directories being added for other application of SpecTf, the `spectf/` directory now serves as the main installable package with an application-agnostic implementation of the model architecture. The primary role of this package is to be imported in code, e.g. `from spectf.models import SimpleSeqClassifier`

The `spectf_cloud` directory then serves as an auxiliary package that provides CLI entry points for calling the trained cloud model for cloud mask generation. It was renamed from `cloud` to avoid collision with other packages, and to start a consistent naming scheme (e.g. `spectf_ch4`). This directory is installed as a package mostly for the script CLI entry point:

https://github.com/emit-sds/SpecTf/blob/2f4b5a5e8490c31ffa7af27451671c5eb80ac084/pyproject.toml#L29

### Test fixes

The class label for cross entropy must be a long int. Single line change, see c44be85c66410619bec2c5a34ebceb80da43fbbc

### Model architecture updates

The model architecture code has been further modularized to support encoder, decoder, and encoder/decoder models, as well as classification and regression models. The model was also renamed from `SimpleSeqClassifier` to `SpecTfEncoder` to make it more generic. All `spectf_cloud` scripts have been updated to use the new model architecture code.

`weights.pt` was restructured to contain the correct `state_dict` keys for this new model architecture definition.

### Pylint and docstrings

The `spectf` package is fully linted and documented. I even adhered to 80 columns. 

## Validation

✅ All tests pass.

```
$ spectf_cloud cloud-eval spectf datasets/lbox_1e4_v6.hdf5 datasets/mmgis_1e4_v6.hdf5 --test-csv datasets/test_fids.csv --weights renamed_weights.pt --gpu 0
Loaded dataset with shape: (3575442, 268)
Test: 100%|███████████████████████████████████| 485/485 [00:26<00:00, 18.20it/s]
Test Accuracy:  0.9526
Test Precision: 0.9604
Test Recall:    0.9436
Test AP: 0.9879
Test AUC: 0.9848
Test TPR: 0.9436
Test FPR: 0.0385
0.52 F1 : 0.9519106009182654
0.52 F05 : 0.9569849180166227
0.52 F025: 0.9593915983903945
0.52 F01 : 0.9602284232063435
Test F1   : 0.9519 @ 0.52
Test F0.5 : 0.9689 @ 0.86
Test F0.25: 0.9829 @ 0.93
Test F0.1 : 0.9932 @ 0.99
```

### CODE REVIEW

File: spectf/model.py [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+        """
+        super().__init__()
+        self.linear = nn.Linear(2, n_filters)
+
+        if activation == 'tanh':
+            self.activation = nn.Tanh()
+        elif activation == 'relu':
+            self.activation = nn.ReLU()
+        elif activation == 'gelu':
+            self.activation = nn.GELU()
```
Comment (michaelkiper) [inline spectf/model.py:91]: Question: Are other activation functions used in the methane work? Using ReLU and GELU for the embeddings doesn't make much intuitive sense to me for an initial embedding given they're not symmetrical

File: spectf/model.py [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+            torch.Tensor: Output tensor of shape (b, s, dim_model)
+        """
+
+        x = self.self_attn(self.norm1(x), self.norm1(x), self.norm1(x))
+        x = self.cross_attn(self.norm2(x), enc_out, enc_out)
+        x = self.ff(self.norm3(x))
+        return x
+
+
+class SpecTfEncoder(nn.Module):
```
Comment (michaelkiper) [inline spectf/model.py:333]: Comment:
My preference would be to either rename this back to `SpecTfClassifier` or create a new class called `SpecTfClassifier` that takes the `SpecTfEncoder` and adds on a classification head (former is preferred). I think that it's a it misleading here are this block does perform classification (`num_classes` param) where encoders just try to project the input into some representational space.

File: spectf/model.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+        """
+        residual = x
+        x = self.linear2(self.dropout(self.gelu(self.linear1(x))))
+        x = self.dropout2(x)
+        if self.use_residual:
+            x = x + residual
+        return x
+
+
+class AttentionBlock(nn.Module):
```
Comment (michaelkiper) [inline spectf/model.py:164]: Praise:
really nice job breaking this out into functional blocks and adding in lots of comments - will help significantly with usability

File: spectf/toa.py [hunk 4]
```diff
@@ -1,34 +1,44 @@
...
+        rdnfp (str): File path to the radiance data (L1b product).
+        obsfp (str): File path to the observation data (L1b product).
+        irrfp (str): File path to the irradiance data (.npy).
 
     Returns:
-    - toa_refl (numpy.ndarray): The calculated top-of-atmosphere reflectance.
-    - banddef (numpy.ndarray): Array of band definitions derived from the radiance data.
-    - metadata (dict): Metadata from the radiance data header.
+        toa_refl (np.ndarray): The calculated top-of-atmosphere reflectance.
+        banddef (np.ndarray): Array of band definitions from .
```
Comment (michaelkiper) [inline spectf/toa.py:29]: Nit:
`Array of band index values.` or something along those lines - sentence is missing the latter part

File: spectf/utils.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        banddef (np.ndarray): The original band definitions.
+        wls (list, optional): The list of wavelengths to be removed.
+                              Default is defined for EMIT.
+    
+    Returns:
+        np.ndarray: The modified band definitions.
+    """
+
+    if wls is None:
+        wls = [381.0055, 388.4092, 395.8158, 403.2254,
```
Comment (michaelkiper) [inline spectf/utils.py:108]: Comment:
Would be nice to break the wavelengths used here into some sort of config file. Made issue to track - https://github.com/emit-sds/SpecTf/issues/24

File: spectf/model.py [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+    
+    This is the simplest Spectral Transformer architecture, consisting only of
+    encoder layers. For most tasks, a single encoder layer is a good starting
+    point. Aside from hyperparameters that affect the dimensionality of layers,
+    the aggregation method is the most significant parameter to tune. Some
+    intuition suggests that 'max' aggregation is appropriate for tasks where
+    a few bands are more important than others, while 'mean' aggregation is
+    appropriate when the overall shape of the spectrum is important. While the
+    'flat' aggregation avoids pooling altogether, it fixes the length of the
+    input sequence, and limits the model to a fixed set of bands.
```
Comment (michaelkiper) [inline spectf/model.py:344]: Optional nit:
Would be nice to also add in a statement about Xavier uniform init for weights and 0 init for biases

File: pyproject.toml [hunk 7]
```diff
@@ -26,8 +26,11 @@ dependencies = [
   "xgboost >= 2.1.3",
 ]
 
-scripts = { spectf = "cloud.cli:main" }
+scripts = { spectf_cloud = "spectf_cloud.cli:main" }
```
Comment (michaelkiper) [inline pyproject.toml:29]: Could you please change this to `scripts = { spectf-cloud = "spectf_cloud.cli:main" }`, that way there isn't a mix of underscores and dashes in the CLI 
Screenshot 2025-01-31 at 12 24 18 PM

**Review by michaelkiper — CHANGES_REQUESTED** [review]
Great job with this refactor! Left a few comments :)

File: spectf/model.py (hunk 1 above)
Comment (jakehlee) [reply spectf/model.py:91]: I just took another look, we did try None but not relu/gelu. I think I agree that we should encourage symmetric functions here, I'll drop relu/gelu.

File: spectf/model.py (hunk 3 above)
Comment (jakehlee) [reply spectf/model.py:164]: Thanks! We definitely needed this update a long time ago

File: spectf/toa.py (hunk 4 above)
Comment (jakehlee) [reply spectf/toa.py:29]: Great catch, fixed.

File: spectf/utils.py (hunk 5 above)
Comment (jakehlee) [reply spectf/utils.py:108]: Agreed, let's track this as a utils constant in the future

File: spectf/model.py (hunk 2 above)
Comment (jakehlee) [reply spectf/model.py:328]: _Technically_ this model doesn't classify because there's no softmax layer at the end - you _could_ set `num_classes=1` and do regression.

Still, I do like the idea of keeping a SpecTf base class and creating specific classification and regression child models with their specific heads. I've created #26 to address this, let's get this merged so this PR unblocks other work.

Comment (michaelkiper) [reply spectf/model.py:328]: Cool - works for me. Thanks for making the issue to track

**Review by michaelkiper — APPROVED** [review]
Great work!
